contrib/jms/src/main/java/com/google/cloud/dataflow/contrib/jms/JmsIO.java

```diff
@@ -0,0 +1,532 @@
+            return new Bound<>(DEFAULT_JMS_CODER).topic(name);
+        }
+
+        /**
+         * Creates and returns a transform for reading from JMS broker that uses the given
+         * {@link Coder} to decode JMS messages into a value of type {@code T}.
+         *
+         * @param coder the coder to use
+         * @return the Bound
+         */
+        public static <T> Bound<T> withCoder(Coder<T> coder) {
+            return new Bound<>(coder);
+        }
+
+        /**
+         * Creates and returns a tranform for reading from JMS broker that limit the
+         * number of messages consumed.
+         *
+         * @param maxNumMessages the max number of messages consumed
+         * @return the Bound
+         */
+        public static Bound<String> maxNumMessages(int maxNumMessages) {
+            return new Bound<>(DEFAULT_JMS_CODER).maxNumMessages(maxNumMessages);
+        }
+
+        /**
+         * A {@link PTransform} that reads from a JMS broker and returns
+         * a unbounded {@link PCollection} containing the items from the stream.
+         */
+        public static class Bound<T> extends PTransform<PInput, PCollection<T>> {
+
+            /** The JMS connection factory. */
+            private final ConnectionFactory connectionFactory;
+
+            /** The JMS queue to read from. */
+            private final String queue;
+
+            /** The JMS topic to read from. */
+            private final String topic;
+
+            /** The coder used to decode each message. */
+            @Nullable
+            private final Coder<T> coder;
+
+            /** Stop after reading this many messages. */
+            private final int maxNumMessages;
+
+            private Bound(Coder<T> coder) {
+                this(null, null, null, null, coder, 0);
+            }
+
+            private Bound(String name, ConnectionFactory connectionFactory, String queue,
+                          String topic, Coder<T> coder, int maxNumMessages) {
+                super(name);
+                this.connectionFactory = connectionFactory;
+                this.queue = queue;
+                this.topic = topic;
+                this.coder = coder;
+                this.maxNumMessages = maxNumMessages;
+            }
+
+            /**
+             * Returns a transform that's like this one but with the given step name.
+             *
+             * @param name the step name
+             * @return the Bound
+             */
+            public Bound<T> named(String name) {
+                return new Bound<T>(name, connectionFactory, queue, topic, coder, maxNumMessages);
+            }
+
+            /**
+             * Returns a transform that's like this one but using the JMS connection factory.
+             *
+             * @param connectionFactory the JMS connection factory to use
+             * @return the Bound
+             */
+            public Bound<T> connectionFactory(ConnectionFactory connectionFactory) {
```

Comment:
(I now assume this is because of using PubSUbIO, which does not follow our most modern conventions.)